### DISCUSSION TITLE
How to use rollup with SvelteKit?

### DISCUSSION THREAD

**Discussion by cayolblake** [sveltejs/kit#6848] [Q&A] [answered]
Hello all,

Is it possible to use rollup with Sveltekit to generate a bundled and stripped down build? Any tutorial there on doing so?

**Comment by teemingc** [comment]
Doesn't Vite already do this?

**Reply by ZetiMente** [reply]
`pnpm run build`

**Comment by cayolblake** [comment]
@s3812497  no it doesn't. You need uglify or terser through rollup to get this done. Which is what I am trying to do with no luck!

**Reply by teemingc** [reply]
It seems that Vite uses rollup to bundle its dependencies. I'm not sure if this is what you're looking for though.

https://vitejs.dev/guide/why.html#why-not-bundle-with-esbuild

https://github.com/vitejs/vite/discussions/7622#discussioncomment-2514757

**Comment by david-plugge** [comment]
Vite uses rollup and does exactly what you want on `npm run build`. Have a look at the generated js files inside `.svelte-kit/output/client/_app/immutable` after executing that command.
If you want to bundle the whole app into a single file have a look at [this comment](https://github.com/sveltejs/kit/issues/2042#issuecomment-889818690)

**Comment by cayolblake** [comment]
@s3812497 @david-plugge 

Vite's rollup does not include treeshaking, or the usage of terser/uglify which removes any JavaScript code that is not used, sometimes called dead code.

In terser for example, `compress: {defaults: false, dead_code: true, unused: true}` which ends up removing both unused and dead code, in addition to https://rollupjs.org/guide/en/#tree-shaking

Which I am stumbling upon how to do exactly in the newly messed up Vite setup which doesn't allow for an external rollup.conf.js where it was easy to do. 🤦🏼‍♂️

**Reply by teemingc** [reply]
Is it possible to pass on your rollup config options to vite?
https://vitejs.dev/config/build-options.html#build-rollupoptions

It also seems that there's some vite compatibility with rollup plugins
https://vitejs.dev/guide/api-plugin.html#rollup-plugin-compatibility

Could this rollup-terser plugin be used? (I've got no experience with rollup or terser)
https://www.npmjs.com/package/rollup-plugin-terser

**Reply by ZetiMente** [reply, marked as answer]
Rich Harris is literally the author of Rollup. You think he would sit by and let Sveltekit not have tree shaking ? 

You can even use terse if you want but it's not even the best option. 

![Screen Shot 2022-09-23 at 11 27 35 PM](https://user-images.githubusercontent.com/76985631/192079864-17acdaef-07bd-4ce0-b26f-5781c898c1c1.png)
